src/GovUk.Education.ExploreEducationStatistics.Admin/Services/FileStorageService.cs

```diff
@@ -49,58 +50,110 @@ public async Task<IEnumerable<FileInfo>> ListFilesAsync(Guid releaseId, ReleaseF
             return await ListFilesAsync(releaseId.ToString(), type);
         }
 
-        public async Task<Either<ValidationResult, IEnumerable<FileInfo>>> UploadDataFilesAsync(Guid releaseId, IFormFile dataFile, IFormFile metaFile, string name)
+        public async Task<Either<ValidationResult, IEnumerable<FileInfo>>> UploadDataFilesAsync(Guid releaseId,
+            IFormFile dataFile, IFormFile metadataFile, string name)
         {
             var blobContainer = await GetCloudBlobContainer();
-            var dataInfo = new Dictionary<string, string> {{NameKey, name}, {MetaFileKey, metaFile.FileName}};
+            var dataInfo = new Dictionary<string, string> {{NameKey, name}, {MetaFileKey, metadataFile.FileName}};
             var metaDataInfo = new Dictionary<string, string> {{DataFileKey, dataFile.FileName}};
-            return await UploadFileAsync(blobContainer, releaseId, dataFile, ReleaseFileTypes.Data, dataInfo)
-                .OnSuccess(async () =>
-                    await UploadFileAsync(blobContainer, releaseId, metaFile, ReleaseFileTypes.Data, metaDataInfo))
-                .OnSuccess(async () =>
-                    await ListFilesAsync(releaseId, ReleaseFileTypes.Data));
+            return await ValidateDataFilesForUpload(blobContainer, releaseId, dataFile, metadataFile)
+                .OnSuccess(async () => await UploadFileAsync(blobContainer, releaseId, dataFile, ReleaseFileTypes.Data, dataInfo))
```

Comment:
Do you actually need the async and awaits on each of these lines, if each of the calls are chained?

Reply:
In order to get it to compile - yes... The issue here I think is that all of the methods are async and there is plenty of magic converting of types to convert signatures. If that sounds a bit waflly then it is - but this was the only way I could get it to work when doing it with async stuff. I was easier when it was synchronous

Reply:
actually now you say it - it might work I'll have a look